### PR TITLE
Use batch validation for Sapling proofs and signatures

### PR DESCRIPTION
**Pull request by nuttycom**
This is based on #6000 

The original branch is str4d/zcash@sapling-batch-validation; opening this PR from my fork as I don't have permissions to push this back to @str4d's remote.

### CODE REVIEW

File: src/rust/src/sapling.rs [hunk 1]
```diff
@@ -174,3 +331,41 @@ impl Verifier {
...
+    Box::new(BatchValidator(Some(sapling_proofs::BatchValidator::new())))
+}
+
+impl BatchValidator {
+    /// Checks the bundle against Sapling-specific consensus rules, and queues its
+    /// authorization for validation.
+    ///
+    /// Returns `false` if the bundle doesn't satisfy the checked consensus rules. This
+    /// `BatchValidator` can continue to be used, but it may include some or all of the
+    /// bundle's proofs and signatures.
```
Comment (daira) [inline src/rust/src/sapling.rs:347]: I don't understand this. What does it mean for a `BatchValidator` to "include" some of the proofs and signatures only when the return value is false? Or is this independent of the return value?

**Review by daira — COMMENTED** [review]
Provisional ACK for f80a61df4fd2c6380a0f574672fdf5d53b89c7da (obviously after splitting into a separate PR) with a question.

**Comment by str4d** [comment]
Rebased to make some improvements now that #6000 and #6046 are merged. This PR will need to be force-pushed at least once more in order to migrate to `zcash_proofs 0.7.1` once that is published (which will be after we have confirmed this PR does what we need).

File: Cargo.toml [hunk 2]
```diff
@@ -92,3 +92,11 @@ features = ["ansi", "env-filter", "fmt", "time"]
...
 panic = 'abort'
 codegen-units = 1
+
+[patch.crates-io]
+f4jumble = { git = "https://github.com/zcash/librustzcash.git", rev = "2519e16ad8ab3a6af59a49028e014e13dddfc36c" }
+zcash_address = { git = "https://github.com/zcash/librustzcash.git", rev = "2519e16ad8ab3a6af59a49028e014e13dddfc36c" }
+zcash_encoding = { git = "https://github.com/zcash/librustzcash.git", rev = "2519e16ad8ab3a6af59a49028e014e13dddfc36c" }
+zcash_note_encryption = { git = "https://github.com/zcash/librustzcash.git", rev = "2519e16ad8ab3a6af59a49028e014e13dddfc36c" }
+zcash_primitives = { git = "https://github.com/zcash/librustzcash.git", rev = "2519e16ad8ab3a6af59a49028e014e13dddfc36c" }
+zcash_proofs = { git = "https://github.com/zcash/librustzcash.git", rev = "2519e16ad8ab3a6af59a49028e014e13dddfc36c" }
```
Comment (daira) [inline Cargo.toml:102]: I checked that this hash matches the current last commit on https://github.com/zcash/librustzcash/pull/580.

**Review by nuttycom — COMMENTED** [review]
testedACK; I can't make this an official github "approval" because I opened the PR, despite the fact that I didn't write it.

File: src/uint256.h [hunk 3]
```diff
@@ -92,7 +92,7 @@ class base_blob
     std::array<uint8_t, WIDTH> GetRawBytes() const
     {
         std::array<uint8_t, WIDTH> buf = {};
-        memcpy(buf.data(), this->begin(), WIDTH);
+        std::memcpy(buf.data(), this->begin(), WIDTH);
```
Comment (daira) [inline src/uint256.h:95]: s/memset/std::memset/ on lines 27 and 42.

**Review by daira — APPROVED** [review]
utACK with suggestions.

File: src/uint256.h (hunk 3 above)
Comment (daira) [reply src/uint256.h:95]: I will file a separate issue or PR for that.

**Comment by str4d** [comment]
[Force-pushed](https://github.com/zcash/zcash/compare/fab9e9b9f0a738ff66b0e4830dd856d235aa6fae..828f566f1c2bbbeed06197c6cabf2752b32c170c) to fix the gtest failure, and add a Sapling-full block benchmark.

Average validation time for block 1723244 on my Ryzen 9 5950X:
- First commit (effectively 5.0.0): 8.94s
- Second commit (with batching): 3.87s
- 
The current PR therefore gives a reduction of 56.7%. Note that both of these are running on a single thread, so there are futher reductions we could get, either by splitting the batch items across multiple batches, or using threadpools inside the batch logic (which @ebfull is currently trying out).

File: Cargo.toml (hunk 2 above)
Comment (daira) [reply Cargo.toml:102]: This is no longer the case. It will be sorted out when we update `zcash_proofs` to depend on bellman 0.13.1.

**Comment by superbaud** [comment]
Yeah, running gitian on https://github.com/nuttycom/zcash/tree/sapling-batch-validation (hash`828f566f1c2bbbeed06197c6cabf2752b32c170c`) failed with

```
then echo Extracting pre-vendored crates from /home/debian/cache/common/vendored-crates.tar.gz...; \
	tar xf /home/debian/cache/common/vendored-crates.tar.gz -C /home/debian/build/zcash/depends/x86_64-linux-gnu; \
else echo Vendoring crates...; \
	/home/debian/build/zcash/depends/x86_64-linux-gnu/native/bin/cargo vendor --locked --manifest-path /home/debian/build/zcash/depends/../Cargo.toml /home/debian/build/zcash/depends/x86_64-linux-gnu/vendored-sources; \
fi
Vendoring crates...
    Updating git repository `https://github.com/zcash/librustzcash.git`
error: failed to sync

Caused by:
  failed to load pkg lockfile

Caused by:
  failed to resolve patches for `https://github.com/rust-lang/crates.io-index`

Caused by:
  failed to load source for dependency `f4jumble`

Caused by:
  Unable to update https://github.com/zcash/librustzcash.git?rev=2519e16ad8ab3a6af59a49028e014e13dddfc36c#2519e16a

Caused by:
  object not found - no match for id (2519e16ad8ab3a6af59a49028e014e13dddfc36c); class=Odb (9); code=NotFound (-3)
make: *** [Makefile:129: /home/debian/build/zcash/depends/x86_64-linux-gnu/.stamp_dce65f034f8] Error 101
make: Leaving directory '/home/debian/build/zcash/depends'
```

and the relevant [cargo lint failed](https://github.com/zcash/zcash/runs/7185185707?check_suite_focus=true):
```
Missing replacement in .cargo/config.offline:
  [source."https://github.com/zcash/librustzcash.git"]
  git = "https://github.com/zcash/librustzcash.git"
  rev = "2519e16ad8ab3a6af59a49028e014e13dddfc36c"
  replace-with = "vendored-sources"
```

**Comment by str4d** [comment]
Re-ran the benchmarks from the first commit in this PR on my Ryzen 9 5950x:
| Unbatched (s) | Batched (s) | Reduction |
|---|---|---|
| 8.88 | 1.88 | 78.8% |

<details>
<summary>Benchmark data</summary>

### Unbatched
```json
[
  {
    "runningtime": 8.884456
  },
  {
    "runningtime": 8.887702000000001
  },
  {
    "runningtime": 8.880763
  },
  {
    "runningtime": 8.86946
  },
  {
    "runningtime": 8.870950000000001
  },
  {
    "runningtime": 8.872097
  },
  {
    "runningtime": 8.880618999999999
  },
  {
    "runningtime": 8.871511999999999
  },
  {
    "runningtime": 8.887917999999999
  },
  {
    "runningtime": 8.873296
  }
]
```

### Batched
```json
[
  {
    "runningtime": 1.900179
  },
  {
    "runningtime": 1.88112
  },
  {
    "runningtime": 1.882656
  },
  {
    "runningtime": 1.888303
  },
  {
    "runningtime": 1.879558
  },
  {
    "runningtime": 1.892048
  },
  {
    "runningtime": 1.882502
  },
  {
    "runningtime": 1.880664
  },
  {
    "runningtime": 1.881228
  },
  {
    "runningtime": 1.880931
  }
]
```
</details>

**Review by ebfull — APPROVED** [review]

**Pull request merged**